quetz/config.py

```diff
@@ -24,207 +25,276 @@
+    allow_credentials: bool = True
+    allow_methods: list[str] = ["*"]
+    allow_headers: list[str] = ["*"]
+
+    class Config:
+        env_prefix = 'cors'
+
+
+class SettingsGitHub(BaseSettings):
+    client_id: str
+    client_secret: str
+
+    class Config:
+        env_prefix = 'github'
+
+
+class SettingsGitLab(BaseSettings):
+    url: str = "https://gitlab.com"
+    client_id: str
+    client_secret: str
+
+    class Config:
+        env_prefix = 'gitlab'
+
+
+class SettingsAzureAD(BaseSettings):
+    client_id: str
+    client_secret: str
+    tenant_id: str
+
+    class Config:
+        env_prefix = 'azuread'
+
+
+class SettingsSQLAlchemy(BaseSettings):
+    database_url: str
+    database_plugin_path: str = ""
+    echo_sql: bool = False
+    postgres_pool_size: int = 100
+    postgres_max_overflow: int = 100
+
+    class Config:
+        env_prefix = 'sqlalchemy'
+
+
+class SettingsSession(BaseSettings):
+    secret: str
+    https_only: bool = True
+
+    class Config:
+        env_prefix = 'session'
+
+
+class SettingsLocalStore(BaseSettings):
+    redirect_enabled: bool = False
+    redirect_endpoint: str = "/files"
+    redirect_secret: str = ""
+    redirect_expiration: int = 3600
+
+    class Config:
+        env_prefix = 'local_store'
+
+
+class SettingsS3(BaseSettings):
+    access_key: str = ""
+    secret_key: str = ""
+    url: str = ""
+    region: str = ""
+    bucket_prefix: str = ""
+    bucket_suffix: str = ""
+
+    class Config:
+        env_prefix = 's3'
+
+
+class SettingsAzureBlob(BaseSettings):
+    account_name: str = ""
+    account_access_key: str = ""
+    conn_str: str = ""
+    container_prefix: str = ""
+    container_suffix: str = ""
+
+    class Config:
+        env_prefix = 'azure_blob'
+
+
+class SettingsGCS(BaseSettings):
+    project: str = ""
+    token: str = ""
+    bucket_prefix: str = ""
+    bucket_suffix: str = ""
+    cache_timeout: int | None = None
+    region: str | None = None
+
+    class Config:
+        env_prefix = 'gcs'
+
+
+class SettingsGoogle(BaseSettings):
+    client_id: str
+    client_secret: str
+
+    class Config:
+        env_prefix = 'google'
+
+
+class SettingsLogging(BaseSettings):
+    level: str = "INFO"
+    file: str = ""
+
+    class Config:
+        env_prefix = 'logging'
 
-    def casted(self, value):
-        if self.cast is bool:
-            try:
-                value = strtobool(str(value))
-            except ValueError as e:
-                raise ConfigError(f"{self.name}: {e}")
 
-        return self.cast(value)
+class SettingsUsers(BaseSettings):
+    admins: list[str] = []
+    maintainers: list[str] = []
+    members: list[str] = []
+    default_role: str = ""
+    collect_emails: bool = False
+    create_default_channel: bool = False
 
+    class Config:
+        env_prefix = 'users'
 
-class ConfigSection(NamedTuple):
-    name: str
-    entries: List[ConfigEntry]
-    required: bool = True
+
+class SettingsWorker:
+    type: str = "thread"
```

Comment:
I don't know which options there are for the type of worker, but I guess it's limited. Perhaps use a `Literal` here with the options? This is rather minor and not critical for this PR.

Reply:
I'm also not sure. In the first pass, I was just copying from the previous config backend. Once this is all working I will convert this to an enum.